### ISSUE TITLE
`vim.g` and others should update "container" types in an expected way.

### ISSUE THREAD

**Issue by tjdevries** [neovim/neovim#12544]
### Actual behaviour

```vim
let g:var = {}
lua vim.g.x = 1
echo g:var
" Result: {}
```

### Expected behaviour

```vim
let g:var = {}
lua vim.g.x = 1
echo g:var
" Result: {'x': 1}
```

It's confusing that it doesn't do this currently I think for people. Either we should do it automatically or find some way to make it "easy" to happen without completely setting/resetting the dict every time.

**Comment by tjdevries** [comment]
@norcalli thoughts? We can use metatables to make this work just fine.

**Comment by norcalli** [comment]
True, I do think it is confusing. My concerns would be:

- Might be too magical. If you keep a reference to a magic dict around, but the variable is overrided to a non dict, then setting that previous reference does... what?
- Dictionaries can contain vim function references, which we currently don't support. I'm actually not sure if g:variables can store functions right now without thinking of the dict case...

The possibilities are:
- fully support vim/lua interop
- support just for dictionaries and ignore incompatibility problems as user error
- raise an error when trying to operate on these returned dicts via metamethods.

Is there any other special type than dictionary and vim functions, btw?

**Comment by mg979** [comment]
@tjdevries did you mean
```vim
let g:var = {}
lua vim.g.var.x = 1
echo g:var
" Result: {'x': 1}
```

Anyway, how can I set/change a nested value in a vim global (dictionary) variable from a lua script, currently? I couldn't find a way.

**Comment by tjdevries** [comment]
Sorry, yeah, that's waht I meant to write @mg979 

The way you could do it now is:

```lua
local t = vim.g.var
t.x = 1
vim.g.var = t
```
that should work, but it is not nice :'(

**Comment by mg979** [comment]
@tjdevries  thanks

**Comment by justinmk** [comment]
Current model is simple. Simple is good. Vimscript-Lua bridge is an escape hatch, it is not something we should pour lots of engineering effort into unless it is blocking a long-term use-case. Vimscript <=> Lua is a short-term workaround, most logic should be in Lua.

Also even after expending a lot of effort we will still have strange edge cases. Whereas with the current simple approach, it is clear and easy to understand the limitations.

> raise an error when trying to operate on these returned dicts via metamethods.

That might be helpful to plugin authors.

**Comment by ipod825** [comment]
I have a somewhat related question.

I have a lua class `Buffer` (table with metatable) corresponding to vim buffer. I'd like to store a reference to a `Buffer` in a vim buffer:
```lua
local lua_buffer = Buffer()
vim.api.nvim_buf_set_var(0, 'lua_buffer', lua_buffer)
```
This way, I don't need to worry about memory leak after the vim buffer is wiped out. 
However, since `lua_buffer` contains meta table and `nvim_buf_set_var` does not seem to support translating meta table (I can see the complexity here), so what is stored into `vim.b.lua_buffer` is only non-meta stuff in `lua_buffer`.

My current workaround is by having a **global** table mapping from vim buffer id to lua `Buffer` and remove `Buffer` from the table on corresponding vim buffer wipeout autocmd.

> Vimscript <=> Lua is a short-term workaround, most logic should be in Lua.

So my question is, is there a long term plan for having lua module replacing vim's scoped variables? For e.g., a table `lb` with same lifecycle as vim buffer.
```lua
vim.lb.var = 'only lives in lua'
assert(vim.b.var==nil)
```

**Comment by justinmk** [comment]
> So my question is, is there a long term plan for having lua module replacing vim's scoped variables? For e.g., a table `lb` with same lifecycle as vim buffer.

@ipod825 you're right that lifetimes (`b:`, `w:`, `t:`) are an important use-case.  `vim.b=function()...` works surprisingly well but you are asking specifically about metatable support. 

Currently, Lua _closures_ seem to work nicely in the Vimscript-Lua bridge. But storing a metatable doesn't do anything useful afaict:

```lua
local f = function()
  local M = {}
  local l = { a=1 }
  local call = function(_, f)
    if l[f] then
      l[f] = l[f] + 1
    else
      l[f] = 1
    end
    print(vim.inspect(l))
  end
  setmetatable(M, {
     __call = call,
     __index = call,
  })
  vim.b.foo = M
  GlobalM = M
  -- vim.api.nvim_buf_set_var(0, 'foo', M)
end
f()

-- Test cases:
--    :lua print(GlobalM['b'])
--        => works
--    :lua print(GlobalM('test'))
--        => works
--    :echo b:foo
--        => []
--    :call b:foo()
--        => E884
```

## Proposal

What if the bridge worked like this for metatables:

1. in Vimscript, it just returns an empty `[]` (seems like it _already_ does this)
    - nothing "bridged" except the variable name
1. such Lua refs live entirely in Lua-land in internal buffer/window/tab tables.
    - when a buffer/window/tab closes, Nvim removes the Lua ref from the relevant Lua table. Lua will garbage-collect it unless some other Lua stuff is still using it.

**Comment by ipod825** [comment]
The proposal looks good to me.  As long as there are complete lua APIs for me to modify nvim internals, I don't really see any usecase to modify something on both sides of the worlds. Using vimscript in lua most of the times is due to lack of lua APIs like augroup/user command pre nvim 0.7.

**Comment by fredizzimo** [comment]
> * Not supported for `g:`/`l:` since there is no use-case?

At least for Neovide there are use cases for `g:`. We would like to have a hierarchy of settings instead of the flat ones currently in use. Initially for more advanced font configuration, and better structure of settings. But there will probably be more use cases in the future. 

It would  also allow us to get rid of the dictwatcher, at the cost of no longer being configurable from the vimscript side. We also publish a few variables that should be read-only by the user, and again a metatable would allow us to enforce that.

The alternative is a `setup` style configuration, like most plugins. But since everything can be configured at runtime, I think it's a poor fit. `vim.g.neovide.*`, would be much nicer. Of course something like `require("neovide").config.value=1` is also possible, but that's not a very standard way of doing things.

I have also seen various proposals for configuring plugins using `vim.g.pluginname.*`, and I think supporting, metatables, or even just updatable dictionaries is a requirement for that.

**Comment by justinmk** [comment]
> It would also allow us to get rid of the dictwatcher, at the cost of no longer being configurable from the vimscript side.

Yep. Good tradeoff imo.

**Comment by lewis6991** [comment]
Have you considered just using `_G` in Lua? I'm not sure why everyone is trying to reach for `vim.g`/`g:` which is implemented in vimscript, and hence will be forever plagued with all the vimscript limitations. I don't see what it offers over `_G`. My guess is that it is a convention people are familiar with?

**Comment by fredizzimo** [comment]
We do have some planned options that will be window specific, so then it would be more consistent to use `vim.g` and `vim.w`, unless we use some meta magic of course.

In the ideal world, it would also be nice to not have to give up the vimscript/commandline support.

**Comment by lewis6991** [comment]
> In the ideal world, it would also be nice to not have to give up the vimscript/commandline support.

For interactive configuration, the current convention is to do everything behind a user command as that'll allow you to process the request without having to watch any variables whilst also allowing you to provide bespoke completion.

**Comment by fredizzimo** [comment]
That would also work, perhaps not as nicely for the user as our current `let g:neovide_setting_name=value`, or with hierarchy, `let g:neovide.setting.name=value` which also has autocompletion and additionally lets you query the current value. And at least in lua could allow assignments of tables. But I guess querying the value could be implemented by using varargs, where no arguments means query.

Commands do have the advantage of making it possible to groups settings that needs to be set at the same time together, so in some cases they are probably better.

We will definitely consider this as an alternative when refacoring our settings, especially since that likely will happen before the metatable support is added and a new Neovim release comes out.

**Comment by ofseed** [comment]
This issue may need to be given more attention.

When I tried to refactor `inlay_hint.lua`, I found that the difficult part of state management (which led to many bugs) was the existence of two states.

1. States related to LSP, such as inlay hints for each client on each buffer.
2. States related to the buffer. Whether inlay hints are enabled on this buffer and which inlay hints have already been applied.

Trying to manage both states simultaneously can be a challenging aspect of state management.

**Comment by justinmk** [comment]
# Proposal

See https://github.com/neovim/neovim/issues/12544#issuecomment-1116794687 (since github loves to hide things).

**Comment by wsdjeg** [comment]
@justinmk Hi, I think the first comment is wrong, in both vim and neovim, we can use `g:` to get all values,  but we can not set `g:` directly.

for example `let g: = {}` cause `E475: Invalid argument: g: = {}`, but we can use `:echo g:` to get all global values.

so the example should be:

```vim
lua vim.g.x = 1

echo g:

" Ruslt: {'x' : 1, 'other_var' : 2}
```
